### PR TITLE
dockerfile written for building docker image

### PR DESCRIPTION
**Pull request by TheMohit2003**
# PR description 
- This pr addresses #360 

## How to test   

1. Ensure Docker is installed on your system.
2. Navigate to the root directory of the project and run:
    ```bash
    docker build -t ruxalab .
    ```
3. Once the image is built, run it using:
    ```bash
    docker run -d --env-file .env -p 5000:5000 ruxalab
    ```
4. Visit `http://localhost:5000` in your browser to view the application.

### CODE REVIEW

**Comment by TheMohit2003** [comment]
@jvJUCA , i have written and tested the docker file . But , haven't update the readme with docker setup steps. And coming to env handling in the project , i have used `--env-file` flag to inject the `env` keys directly in the running container. I will soon send the updated readme on this branch itself .

File: dockerfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
+
+FROM node:16 AS build-stage
```
Comment (asr2003) [inline dockerfile:2]: Please use `node:lts` instead node:16 to mitigate risks of compatibility in future

File: dockerfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+COPY package*.json ./
+
+RUN npm install
+
+COPY . .
+
+RUN npm run build
+
+# Serve the app using Node.js
+FROM node:16 AS production-stage
```
Comment (asr2003) [inline dockerfile:15]: Use` node:alpine` to reduce the size of final image

**Review by asr2003 — COMMENTED** [review]
I would suggest having two docker files for prod and dev environments are helpful rather than a single Dockerfile.

File: dockerfile (hunk 2 above)
Comment (KarinePistili) [reply dockerfile:15]: Hello @asr2003 I think this are good suggestions!

File: dockerfile (hunk 1 above)
Comment (KarinePistili) [reply dockerfile:2]: Good suggestion!

File: dockerfile (hunk 2 above)
Comment (KarinePistili) [reply dockerfile:15]: I think it would good to use node latest here as well

**Comment by KarinePistili** [comment]
@xemyst were you able to test it on your environment as well?

File: dockerfile (hunk 2 above)
Comment (asr2003) [reply dockerfile:15]: Yes, we can use node lts for build stage. node-alpine is the lighter image which we can use safely to run the already built files in Stage-1. It's the mantra we follow to reduce docker image :)

Comment (KarinePistili) [reply dockerfile:15]: Cool!

**Comment by MurilonND** [comment]
Hello @TheMohit2003, I just finished testing your commit, and for me, the code isn't working (tested after merging with the develop so this could be why), so I just changed the code and got it working.

Screenshot from 2024-04-15 15-03-16

**Review by MurilonND — CHANGES_REQUESTED** [review]
Hello @TheMohit2003, I just finished testing your commit, and for me, the code isn't working (tested after merging with the develop so this could be why), so I just changed the code and got it working.

Screenshot from 2024-04-15 15-03-16

**Review by KarinePistili — CHANGES_REQUESTED** [review]
Please review the changes provided by the reviewers and change accordingly

**Comment by TheMohit2003** [comment]
@MurilonND , can you discuss the changes ?

**Comment by MurilonND** [comment]
Sure, what you want to know?

**Comment by TheMohit2003** [comment]
> Sure, what you want to know?

What error did you get, and what file changes did you make ?

**Comment by MurilonND** [comment]
Very simple, just changed the node version to alpine (optimization), and the ERROR that I got was:

[build-stage 6/6] RUN npm run build:
0.364 npm ERR! Missing script: "build"
0.364 npm ERR! 
0.364 npm ERR! To see a list of scripts, run:
0.365 npm ERR!   npm run
0.365 
0.365 npm ERR! A complete log of this run can be found in:
0.366 npm ERR!     /root/.npm/_logs/2024-04-15T17_30_45_835Z-debug-0.log
------
dockerfile:12
--------------------
  10 |     COPY . .
  11 |
  12 | >>> RUN npm run build
  13 |
  14 |     # Serve the app using Node.js
--------------------
ERROR: failed to solve: process "/bin/sh -c npm run build" did not complete successfully: exit code: 1

The develop doesn't have a "build" script, instead, it has "build-dev" or "build-prod."

**Comment by KarinePistili** [comment]
@TheMohit2003 will you be able to do the changes?

**Comment by TheMohit2003** [comment]
@KarinePistili , My apologies for being late. I will send the commits by tomorrow. Just went through distro hopping ( linux )and need to setup stuff again.

**Comment by KarinePistili** [comment]
Okay, thanks! @TheMohit2003

**Comment by TheMohit2003** [comment]
**Updated Dockerfile:**
- Changed the Node.js version in the **build stage** to `node:lts` to ensure long-term support and compatibility.
- Switched to `node:alpine` in the **production stage** to reduce the final image size, improving deployment efficiency.
@KarinePistili

**Comment by TheMohit2003** [comment]
@KarinePistili , should I add docker image setup docs in readme ?

**Comment by KarinePistili** [comment]
it would be nice @TheMohit2003 , thanks

**Comment by TheMohit2003** [comment]
@KarinePistili, all tasks on my end are completed. If there are any further changes you need, please let me know. Otherwise, you are free to merge the PR when you are ready.

**Comment by asr2003** [comment]
I will test your changes and provide any changes needed

**Comment by TheMohit2003** [comment]
@asr2003 , did you reviewed the changes ?

**Comment by JulioManoel** [comment]
I think it's interesting to put a dockerignore for common-models

**Comment by TheMohit2003** [comment]
@JulioManoel worked for me . Let me check again. Did you create a .env file with all the secrets before running the container?

**Comment by JulioManoel** [comment]
Wouldn't it be better to separate two dockerfiles, 1 for develop and another for production?

**Comment by JulioManoel** [comment]
@TheMohit2003 there was an error when building the container

**Comment by JulioManoel** [comment]
I have the .env, I gave npm i to the project and after that I tried to build docker using the README command:

```bash
docker build -t uxremotelab .
```

**Comment by TheMohit2003** [comment]
@JulioManoel , I can do it that way, but the assigned reviewers haven't said anything regarding the current changes . Let me take their reviews in consideration, and then I will write a separate dev and production based dockerfile . @KarinePistili @asr2003 .

**Comment by JulioManoel** [comment]
Sorry it was an internet problem ksks 😅

**Review by JulioManoel — APPROVED** [review]

**Comment by KarinePistili** [comment]
Thank you for revision @JulioManoel @MurilonND @asr2003 and @xemyst

**Pull request merged**